src/code/z_play_hireso.c

```diff
@@ -1,27 +1,1349 @@
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_play_hireso/BombersNotebook_Destroy.s")
+#define DEFINE_EVENT(_enum, _icon, colorFlag, _description, _completedMessage, _completedFlag) colorFlag,
+
+u16 sBombersNotebookEventColorWeekEventFlags[] = {
+#include "tables/bombers_notebook/event_table.h"
+};
+
+#undef DEFINE_EVENT
+
+void BombersNotebook_TextureRectangle(Gfx** gfxP, s32 rxl, s32 ryl, s32 rxh, s32 ryh, s32 tile, s32 s, s32 t, s32 dsdx,
+                                      s32 dtdy) {
+    s32 xl = rxl - (D_801FBBD0 * 4);
+    s32 yl = ryl - (D_801FBBD2 * 4);
+    s32 xh = rxh - (D_801FBBD0 * 4);
+    s32 yh = ryh - (D_801FBBD2 * 4);
+    Gfx* gfx = *gfxP;
+
+    gSPScisTextureRectangle(gfx++, xl, yl, xh, yh, tile, s, t, dsdx, dtdy);
+
+    *gfxP = gfx;
+}
+
+s16 sBombersNotebookHeaderColors[][3] = {
+    { 116, 134, 146 },
+    { 158, 156, 131 },
+    { 174, 141, 151 },
+};
+
+void BombersNotebook_DrawHeaders(Gfx** gfxP) {
+    Gfx* gfx = *gfxP;
+    s32 i;
+    s32 rectLeft;
+
+    gDPLoadTextureBlock(gfx++, gBombersNotebookHeaderBoxTex, G_IM_FMT_I, G_IM_SIZ_8b, 8, 24, 0,
+                        G_TX_NOMIRROR | G_TX_WRAP, G_TX_NOMIRROR | G_TX_WRAP, 3, G_TX_NOMASK, G_TX_NOLOD, G_TX_NOLOD);
+    for (i = 0, rectLeft = 120; i < ARRAY_COUNT(sBombersNotebookHeaderColors); i++, rectLeft += 150) {
+        gDPSetPrimColor(gfx++, 0, 0, sBombersNotebookHeaderColors[i][0], sBombersNotebookHeaderColors[i][1],
+                        sBombersNotebookHeaderColors[i][2], 192);
+        BombersNotebook_TextureRectangle(&gfx, rectLeft * 4, 74 * 4, (rectLeft + 143) * 4, 98 * 4, 0, 0, 0, 0x400,
+                                         0x400);
+    }
+
+    gDPSetPrimColor(gfx++, 0, 0, 255, 255, 255, 255);
+    for (i = 0, rectLeft = 124; i < ARRAY_COUNT(sBombersNotebookDayTextures); i++, rectLeft += 150) {
+        gDPLoadTextureBlock(gfx++, sBombersNotebookDayTextures[i], G_IM_FMT_IA, G_IM_SIZ_8b, 48, 22, 0,
+                            G_TX_NOMIRROR | G_TX_WRAP, G_TX_NOMIRROR | G_TX_WRAP, G_TX_NOMASK, G_TX_NOMASK, G_TX_NOLOD,
+                            G_TX_NOLOD);
+        BombersNotebook_TextureRectangle(&gfx, rectLeft * 4, 77 * 4, (rectLeft + 48) * 4, 99 * 4, 0, 0, 0, 0x400,
+                                         0x400);
+    }
+
+    rectLeft = 172;
+    gDPLoadTextureBlock_4b(gfx++, gBombersNotebook1800Tex, G_IM_FMT_IA, 48, 11, 0, G_TX_NOMIRROR | G_TX_WRAP,
+                           G_TX_NOMIRROR | G_TX_WRAP, G_TX_NOMASK, G_TX_NOMASK, G_TX_NOLOD, G_TX_NOLOD);
+    for (i = 0; i < 3; i++) {
+        BombersNotebook_TextureRectangle(&gfx, rectLeft * 4, 86 * 4, (rectLeft + 48) * 4, 97 * 4, 0, 0, 0, 0x400,
+                                         0x400);
+        rectLeft += 150;
+    }
+
+    gDPPipeSync(gfx++);
+    *gfxP = gfx;
+}
+
+s16 sBombersNotebookColumnColors[][2][3] = {
+    { { 165, 183, 195 }, { 140, 158, 170 } },
+    { { 197, 195, 172 }, { 172, 170, 147 } },
+    { { 223, 190, 200 }, { 190, 165, 175 } },
+};
```

Comment:
The third dimension here feels unnecessary? Could you post a scratch of `BombersNotebook_DrawColumns`, is it really that bad?

Reply:
https://decomp.me/scratch/kZ41w
The function more or less treats it as a single 1D array of s16s (notice the iterator is just a pointer to an s16, and it has to be updated by 3 in the loop. I made it 3D to be clearer on how it is used. The first dimension is which column (day) it is. The second corresponds to the subcolumns (if you look at the notebook each day's background is subdivided into 4, and uses two different colors). The last dimension is the usual RGB colors.

Reply:
Oh I think I see. Yeah that makes sense, could you possibly explain that above the data?